File: python/puzzles/day-11/solution.py
import sys
import os
from collections import defaultdict

# Modifying Path to include Repo Directory (for util import)
sys.path.append(os.path.dirname(os.path.dirname(os.path.dirname(os.path.realpath(__file__)))))
# Import Utilities
import utils

def puzzle(filename, part2):
    # Zero the Accumulator
    score = 0
    
    # Open File
    with open(filename, 'r') as fp:
        # Split the stones out
        stones = defaultdict(int)
        for stone in [int(i) for i in fp.readline().strip().split()]:
            stones[stone] += 1
        
    num = 75 if part2 else 25
    for i in range(num):
        stones = runBlink(stones)
    
    score = sum(stones.values())
    
    # Return Accumulator    
    print(score)
  

# Stones the number of each stone type in a dict
def runBlink(stones):
    newStones = defaultdict(int)
    
    for stone, num in stones.items():
        stoneAsStr = str(stone)
        lenStone = len(stoneAsStr)
        
        if stone == 0:
            newStones[1] += num
        elif lenStone % 2 == 0:
            midpoint = lenStone // 2
            newStones[int(stoneAsStr[midpoint:])] += num
            newStones[int(stoneAsStr[:midpoint])] += num
        else:
            newStones[stone*2024] += num
    
    return newStones
    
if __name__ == "__main__":
    # Check number of Arguments, expect 2 (after script itself)
    # Args: input file, part number
    if (len(sys.argv) != 3):
        print(f"Error: Expected 2 arguments, got {len(sys.argv) - 1}")
        
    else:
        puzzle(sys.argv[1], sys.argv[2] == '2')

File: cpp/puzzles/day-11/solution.cpp
#include <cstddef>
#include <iostream>
#include <fstream>
#include <numeric>
#include <string>
#include <map>
#include "utils.hpp"
using namespace std;

map<long, long> runBlink(map<long, long> stones) {
    map<long, long> newStones;

    for (const auto& pair : stones) {
        long stone = pair.first;
        long num = pair.second;

        string stoneAsStr = to_string(stone);
        int lenStone = stoneAsStr.length();

        if (stone == 0) {
            newStones[1] += num;
        } else if ((lenStone % 2) == 0){
            int midpoint = lenStone / 2;
            newStones[stoi(stoneAsStr.substr(0, midpoint))] += num;
            newStones[stoi(stoneAsStr.substr(midpoint))] += num;
        } else {
            newStones[stone * 2024] += num;
        }
    }

    return newStones;
}

void puzzle(char *input, bool part2) {
    ifstream fp(input, ios_base::in);
    string line;

    // Stones are stored as a frequency map
    map<long, long> stones;

    // Loop over file, line by line
    while (getline (fp, line)) {
        for (string str : splitString(line, ' ')) {
            stones[stoi(str)]++;
        }
    }

    int num = (part2) ? 75 : 25;

    for (int i = 0; i < num; i++) {
        stones = runBlink(stones);
    }

    long score = 0;
    for (map<long, long>::iterator it = stones.begin(); it != stones.end(); ++it) {
        score += it->second;
    }

    cout << score << endl;
}

int main(int argc, char *argv[]) {
    // Expect 2 arguments beyond the caller: input file and part number
    if (argc == 3) {
        puzzle(argv[1], argv[2][0] == '2');
    } else {
        cerr << "Error: Expected 2 arguments, got: " << (argc - 1) << endl;
    }

    return 0;
}
